### PR TITLE
Design document for SSH keys in Node resources

### PR DESCRIPTION
**Pull request by stefwalter**
Kubernetes clusters are often deployment and upgraded over SSH. This
may include the node operating system, the kublet, and so on. SSH tunnels
are also optionally used for node/proxy communication.

This SSH access information should be tracked and configured along with other
Node information, to allow concepts such as self-registering nodes, clusters
with heterogenous Nodes, common deployment and cluster upgrade code, and so on.

---

This change is [Reviewable](http://reviewable.k8s.io/reviews/kubernetes/kubernetes/23811)

### CODE REVIEW

**Comment by stefwalter** [comment]
Some alternatives, and failure modes, evaluated here: https://github.com/cockpit-project/cockpit/wiki/Kubernetes-Exec-on-Node

**Comment by stefwalter** [comment]
@philips You might be interested in thist. Related to what you were talking with monokube.

**Comment by errordeveloper** [comment]
In theory, it's not unreasonable to argue that SSH is just an implementation detail and doesn't provide the most portable way for tasks such as cluster upgrades. Fully-containerised cluster deployment, whether self-hosted or not, shouldn't require SSH access...

**Comment by errordeveloper** [comment]
In practice, it's probably much better to accept the current implementation and provide a thin set of abstractions over it, given it doesn't require enormous effort.

**Comment by errordeveloper** [comment]
First question that arises, is what user should I login as, if I want to use the key I obtained from the API? Secondly, there many things that could go wrong... What comes to mind first is a read-only filesystem, or basic environment variables, such as `PATH` and `LANG`, or even login shell... Also, if a key is absent, does it mean there is no SSH access or some kind of "default" key can be used? Perhaps an explicit API endpoint/attribute should exist to declare whether SSH is at all enabled or not.

**Comment by stefwalter** [comment]
> In theory, it's not unreasonable to argue that SSH is just an implementation detail and doesn't provide the most portable way for tasks such as cluster upgrades. Fully-containerised cluster deployment, whether self-hosted or not, shouldn't require SSH access...

Indeed. Using/driving a kubernetes cluster certainly doesn't require SSH access to Nodes. However there are many cases where this is done today ... including Kubernetes itself with SSH proxy tunnels but also monokube, Commissaire, Ansible and so on. These are places where folks are using SSH to bootstrap, configure and upgrade the cluster. It's the base level of access for many cloud instances, so I guess it follows from there.

> In practice, it's probably much better to accept the current implementation and provide a thin set of abstractions over it, given it doesn't require enormous effort.

The proof of concept mostly does that: Using metadata.annotations. The goal of this design document is that we don't end up with multiple fragmented incompatible abstractions for the same thing.

> First question that arises, is what user should I login as, if I want to use the key I obtained from the API? 

Right, the login isn't consistent between varying instances. So that's a good point. Lets add that (probably to NodeSystemInfo). Although for anyone following along, this isn't about users but rather giving tools/scripts root node access.

> Secondly, there many things that could go wrong... What comes to mind first is a read-only filesystem, or basic environment variables, such as PATH and LANG, or even login shell... Also, if a key is absent, does it mean there is no SSH access or some kind of "default" key can be used? Perhaps an explicit endpoint/attribute should exist to declare whether SSH is at all enabled/allowed or not.

It's not assumed that any node anywhere will want to or be able to expose these facilities. Tools with write access to the Node resource could place a key into the spec ... so there should be some indication that this is enabled/allowed. Will try to work that in.

**Comment by smarterclayton** [comment]
@kubernetes/sig-node needs some discussion

File: docs/design/node-authorized-keys.md [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+Resource outline:
+
+```
+type NodeSpec struct {
+	...
+	NodeAuthorizedKeys []NodePublicKey `json:"nodeAuthorizedKeys",omitempty`
+	...
+}
+
+type NodePublicKey struct {
```
Comment (timstclair) [inline docs/design/node-authorized-keys.md]: I think secrets already provide this, and the fewer security-sensitive abstractions we need the better.

**Comment by timstclair** [comment]
What about using a daemon set to watch for new keys (or deletions) and sync them to the host? (With a hostpath volume). That way you use existing abstractions but can still SSH if the kubelet or container runtime goes down (assuming sshd is still up and reachable).

File: docs/design/node-authorized-keys.md (hunk 1 above)
Comment (smarterclayton) [inline docs/design/node-authorized-keys.md]: Authorized keys aren't technically secret, but I somewhat agree a
DaemonSet sync might be a better place to start (and use a configmap /
secret).

**Comment by ashcrow** [comment]
Daemon Sets might work fine, but part of the want for this change is to have a central location to snag keys when needed. It sounds like secrets mapped to a node might work.

File: docs/design/node-authorized-keys.md (hunk 1 above)
Comment (timstclair) [inline docs/design/node-authorized-keys.md]: Do ConfigMaps have the same level of access control as secrets? The keys should be public keys, but write access needs to be restricted.

Comment (smarterclayton) [inline docs/design/node-authorized-keys.md]: We don't have per resource controls, so these "node config maps" would
have to go in a special namespace.  Of Course, we  want that for other
reasons as well for node level config objects anyway

Comment (stefwalter) [inline docs/design/node-authorized-keys.md]: @smarterclayton A daemon set that configures SSH appropriately? Interesting idea. Will think about that.

**Comment by ashcrow** [comment]
bueller

**Comment by euank** [comment]
I personally think that this proposal does not justify the need for itself.

To take the provided reasons:

> to allow concepts such as self-registering nodes, clusters
with heterogenous Nodes, common deployment and cluster upgrade code, and so on.

* self-registering nodes - The kubelet bootstrap story is an existing effort that exists unrelated to ssh. The kubelet contacts the api-service via TLS, and ssh is not needed in that flow
* heterogenous nodes - Works with or without this, and right now nodes don't even need sshd on them
* cluster upgrade - The self-hosted kubelet stuff uses kubernetes to upgrade kubernetes. We have an existing channel for running commands on a node (api-service -> kubelet -> pod -> commands) which can be used to run a new version of the kubelet, a new version of the api-service, etc.

I do think that some people will want the convenience of cluster-wide ssh keys and access, but I think that should be implemented on top of K8s, not as a part of it (e.g. with secrets and daemonsets as suggested above).

ssh can be used for cluster deployment in many cases, such as via ansible or kube-up scripts, but I don't think any of the kubernetes components themselves should mandate ssh for anything.

**Comment by stefwalter** [comment]
> I do think that some people will want the convenience of cluster-wide ssh keys and access, but I think that should be implemented on top of K8s, not as a part of it (e.g. with secrets and daemonsets as suggested above).

Agree. @petervo is working on such DaemonSets.

> ssh can be used for cluster deployment in many cases, such as via ansible or kube-up scripts, but I don't think any of the kubernetes components themselves should mandate ssh for anything.

Kubernetes already has a mode where it makes active use of SSH to nodes. See: http://kubernetes.io/docs/admin/master-node-communication/

But in any case, as demonstrated by the proof of concept, there's nothing this proposal is blocking. Use of annotations work just as well to decorate Node objects. 

So lets close this. At least until we have multiple implementations that do the same thing, at which point we can revisit this in the name of consistency.

**Pull request closed without merging**